### PR TITLE
MOTECH-2341: Added Update Person Task Action to OpenMRS module.

### PR DESCRIPTION
**Pull request by pkornowski**
Added Update person task action. We can update given name, middle name, family name, gender and address.  Person is searching by UUID. I checked this, when task is invoked, then changes are visible on the openMRS server.

### CODE REVIEW

**Comment by kmadej** [comment]
Now after discussion on JIRA, everything should be clear. However please wait for changes in Person's address.

**Comment by pkornowski** [comment]
- I split one updatePerson method in OpenMRSPersonServiceImpl.java on parts which updates names, addresses and general information, because of when I used updatePerson function before this, it added new address, and names. When I using methods: updatePersonAddress and updatePersonNames it actually works, I think.
- Referring to the discussion: https://applab.atlassian.net/browse/MOTECH-2341 , I tested this and it is possible to update only one field. We have to add only 'Person UUID' field.

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/PersonResourceImpl.java [hunk 1]
```diff
@@ -67,7 +67,7 @@ public void updatePersonName(Config config, String uuid, Person.Name name) {
     @Override
     public void updatePersonAddress(Config config, String uuid, Person.Address address) {
         String requestJson = buildGson().toJson(address);
-        postWithEmptyResponseBody(config, requestJson, "/person/{personUuid}/address/{addressUuid}", uuid, address.getUuid());
+        postWithEmptyResponseBody(config, requestJson, "/person/{personUuid}/address/{addressUuid}?v=full", uuid, address.getUuid());
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/PersonResourceImpl.java]: is this required ? This is POST request with empty response body. I suppose that this 'full' option refers to response body only.

File: openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java [hunk 2]
```diff
@@ -93,7 +93,20 @@ public Person createPerson(String configName, String givenName, String familyNam
     public Person updatePerson(String configName, Person person) {
         try {
             Config config = configService.getConfigByName(configName);
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java:96]: we should add validation for person. I suppose that only uuid is required here, am I right ?

File: openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java [hunk 3]
```diff
@@ -93,7 +93,20 @@ public Person createPerson(String configName, String givenName, String familyNam
     public Person updatePerson(String configName, Person person) {
         try {
             Config config = configService.getConfigByName(configName);
-            Person updated = personResource.updatePerson(config, person);
+            Person fetchedPerson = personResource.getPersonById(config, person.getUuid());
+
+            Person.Address addressForUpdate = fetchedPerson.getPreferredAddress();
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java:101]: Please write a comment why you have to update address and name separately

File: openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java [hunk 4]
```diff
@@ -93,7 +93,20 @@ public Person createPerson(String configName, String givenName, String familyNam
...
+            if(addressForUpdate != null) {
+                person.getPreferredAddress().setUuid(addressForUpdate.getUuid());
+            }
+            personResource.updatePersonAddress(config, person.getUuid(), person.getPreferredAddress());
+
+            Person.Name nameForUpdate = fetchedPerson.getPreferredName();
+            person.getPreferredName().setUuid(nameForUpdate.getUuid());
+            personResource.updatePersonName(config, person.getUuid(), person.getPreferredName());
+
+            Person generalInformationForUpdate = prepareGeneralInformationForUpdate(person);
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java]: As far as I understand, the name and address values should not be sent in update person request. The better/more readable way to do this is using PersonUpdateSerializer

File: openmrs-19/src/main/resources/task-channel.json [hunk 5]
```diff
@@ -182,6 +182,140 @@
...
+    },
+    {
+      "displayName": "openMRS.updatePerson",
+      "serviceInterface": "org.motechproject.openmrs19.tasks.OpenMRSActionProxyService",
+      "serviceMethod": "updatePerson",
+      "actionParameters": [
+        {
+          "key": "personUuid",
+          "displayName": "openMRS.person.uuid",
+          "required": false
```
Comment (kmadej) [inline openmrs-19/src/main/resources/task-channel.json]: this field should be required

File: openmrs-19/src/test/java/org/motechproject/openmrs19/it/MRSPersonServiceIT.java [hunk 6]
```diff
@@ -120,15 +142,30 @@ public void shouldUpdatePerson() throws OpenMRSException, InterruptedException {
         assertEquals(newFirstName, updated.getPreferredName().getGivenName());
         assertEquals(newMiddleName, updated.getPreferredName().getMiddleName());
         assertEquals(newLastName, updated.getPreferredName().getFamilyName());
+
         // So far OpenMRS module stores only one field of createdPerson's address, which is 'address1'.
```
Comment (kmadej) [inline openmrs-19/src/test/java/org/motechproject/openmrs19/it/MRSPersonServiceIT.java]: now OpenMRS module stores many fields of address, so you can adjust this test. Just assert address objects. There is a ticket to create ITs for TaskDataProvider, so then we will check if fullAddressString works correctly

File: openmrs-19/src/test/java/org/motechproject/openmrs19/resource/impl/PersonResourceImplTest.java [hunk 7]
```diff
@@ -158,7 +158,7 @@ public void shouldUpdatePersonName() throws Exception {
     public void shouldUpdatePersonAddress() throws Exception {
         String personId = "CCC";
         Person.Address preferredAddress = prepareAddress();
-        URI url = config.toInstancePathWithParams("/person/{personUuid}/address/{addressUuid}", personId,
+        URI url = config.toInstancePathWithParams("/person/{personUuid}/address/{addressUuid}?v=full", personId,
```
Comment (kmadej) [inline openmrs-19/src/test/java/org/motechproject/openmrs19/resource/impl/PersonResourceImplTest.java]: remove this

File: openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java [hunk 8]
```diff
@@ -93,7 +93,20 @@ public Person createPerson(String configName, String givenName, String familyNam
...
-            Person updated = personResource.updatePerson(config, person);
+            Person fetchedPerson = personResource.getPersonById(config, person.getUuid());
+
+            Person.Address addressForUpdate = fetchedPerson.getPreferredAddress();
+            if(addressForUpdate != null) {
+                person.getPreferredAddress().setUuid(addressForUpdate.getUuid());
+            }
+            personResource.updatePersonAddress(config, person.getUuid(), person.getPreferredAddress());
+
+            Person.Name nameForUpdate = fetchedPerson.getPreferredName();
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java:107]: what about causeOfDeath ? I suppose that it also have to be updated separately, however please check it

File: openmrs-19/src/test/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceTest.java [hunk 9]
```diff
@@ -202,6 +209,24 @@ public void shouldNotUsedDefaultLocationWhenLocationForGivenNameIsNotFound() {
...
+        Person.Address personAddress = person.getPreferredAddress();
+
+        openMRSActionProxyService.updatePerson(person.getUuid(), person.getPreferredName().getGivenName(), person.getPreferredName().getMiddleName(),
+                person.getPreferredName().getFamilyName(), personAddress.getAddress1(), personAddress.getAddress2(),
+                personAddress.getAddress3(), personAddress.getAddress4(), personAddress.getAddress5(),
+                personAddress.getAddress6(), personAddress.getCityVillage(), personAddress.getStateProvince(),
+                personAddress.getCountry(), personAddress.getPostalCode(), personAddress.getCountyDistrict(),
+                personAddress.getLatitude(), personAddress.getLongitude(), new DateTime(personAddress.getStartDate()),
+                new DateTime(personAddress.getEndDate()), new DateTime(person.getBirthdate()), person.getBirthdateEstimated(),
+                person.getGender(), person.getDead(), "");
```
Comment (kmadej) [inline openmrs-19/src/test/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceTest.java]: why causeOfDeath is empty ?

File: openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java (hunk 8 above)
Comment (pkornowski) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java:107]: I checked it, and causeOfDeath field updating currently worked.

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/PersonResourceImpl.java (hunk 1 above)
Comment (pkornowski) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/PersonResourceImpl.java]: After your comment I checked this and in fact it is not needed.

File: openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java (hunk 2 above)
Comment (pkornowski) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java:96]: Yes there is required only person uuid for get person from the OpenMRS server.

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/PersonResourceImpl.java [hunk 10]
```diff
@@ -53,8 +53,8 @@ public void deleteAttribute(Config config, String uuid, Attribute attribute) {
 
     @Override
     public Person updatePerson(Config config, Person person) {
-        String requestJson = buildGsonWithConceptAdapter().toJson(person);
-        String responseJson = postForJson(config, requestJson, "/person/{uuid}?v=full", person.getUuid());
+        String requestJson = buildGsonWithUpdatePersonAdapter().toJson(person);
+        String responseJson = postForJson(config, requestJson, "/person/{uuid}", person.getUuid());
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/PersonResourceImpl.java]: why you remove this 'full' option ?

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/PersonResourceImpl.java [hunk 11]
```diff
@@ -83,6 +83,13 @@ private Gson buildGsonWithConceptAdapter() {
                 .create();
     }
 
+    private Gson buildGsonWithUpdatePersonAdapter() {
+        return new GsonBuilder()
+                .excludeFieldsWithoutExposeAnnotation()
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/PersonResourceImpl.java]: I suppose we do not need this now after using serializer

File: openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java [hunk 12]
```diff
@@ -91,8 +91,26 @@ public Person createPerson(String configName, String givenName, String familyNam
 
     @Override
     public Person updatePerson(String configName, Person person) {
+
         try {
             Config config = configService.getConfigByName(configName);
+
+            Validate.notNull(person.getUuid(), "Person uuid cannot be null");
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java]: validation should be at the beginning

Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java]: use validate.nonempty for string

File: openmrs-19/src/test/java/org/motechproject/openmrs19/it/MRSPersonServiceIT.java [hunk 13]
```diff
@@ -120,15 +142,27 @@ public void shouldUpdatePerson() throws OpenMRSException, InterruptedException {
...
-        // So far OpenMRS module stores only one field of createdPerson's address, which is 'address1'.
-        // However while retrieving createdPerson from OpenMRS server all createdPerson's address fields are put
-        // into one string. That's why it is checked if address field contains address1 value.
-        assertTrue(updated.getPreferredAddress().getFullAddressString().contains(newAddress));
-        assertEquals(newGender, updated.getGender());
 
-        assertTrue(mrsListener.created);
-        assertTrue(mrsListener.updated);
-        assertFalse(mrsListener.deleted);
+        assertEquals(newAddress1, updated.getPreferredAddress().getAddress1());
```
Comment (kmadej) [inline openmrs-19/src/test/java/org/motechproject/openmrs19/it/MRSPersonServiceIT.java]: assert two address objects

File: openmrs-19/src/test/java/org/motechproject/openmrs19/resource/impl/PersonResourceImplTest.java [hunk 14]
```diff
@@ -75,7 +75,7 @@ public void shouldCreatePerson() throws Exception {
     @Test
     public void shouldUpdatePerson() throws Exception {
         Person person = preparePerson();
-        URI url = config.toInstancePathWithParams("/person/{uuid}?v=full", person.getUuid());
+        URI url = config.toInstancePathWithParams("/person/{uuid}", person.getUuid());
```
Comment (kmadej) [inline openmrs-19/src/test/java/org/motechproject/openmrs19/resource/impl/PersonResourceImplTest.java]: why you remove this ?

File: openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java [hunk 15]
```diff
@@ -91,8 +91,25 @@ public Person createPerson(String configName, String givenName, String familyNam
 
     @Override
     public Person updatePerson(String configName, Person person) {
+        Validate.notEmpty(person.getUuid(), "Person uuid cannot be null");
         try {
             Config config = configService.getConfigByName(configName);
+
+            Person fetchedPerson = personResource.getPersonById(config, person.getUuid());
+
+            //Update is splited on separately update address, names and general information, because of
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/service/impl/OpenMRSPersonServiceImpl.java]: please correct this comment. It should be 'Updating address and name of person must be done separately'

**Pull request merged**